eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Handler.php

```diff
@@ -628,46 +639,122 @@ public function locationSwapped($location1Id, $location1ParentId, $location2Id,
+                        $name1,
+                        $language->id,
+                        $isMainLanguage2 && $alwaysAvailable1,
+                        $isMainLanguage2,
+                        $autoLocation1['id']
+                    );
+                } else {
+                    $this->internalPublishUrlAliasForLocation(
+                        $location1Id,
+                        $location1ParentId,
+                        $name1,
+                        $language->id,
+                        $isMainLanguage2 && $alwaysAvailable1,
+                        $isMainLanguage2,
+                        $autoLocation1['id']
+                    );
 
-            foreach ($languageIds as $languageId) {
-                $isMainLanguage = $languageId == $location2MainLanguageId;
+                    $this->internalPublishUrlAliasForLocation(
+                        $location2Id,
+                        $location2ParentId,
+                        $name2,
+                        $language->id,
+                        $isMainLanguage1 && $alwaysAvailable2,
+                        $isMainLanguage1,
+                        $autoLocation2['id']
+                    );
+                }
+            } elseif (isset($namesByLanguage1[$languageCode])) {
                 $this->internalPublishUrlAliasForLocation(
                     $location1Id,
                     $location1ParentId,
-                    $row['text'],
-                    $languageId,
-                    $isMainLanguage && $alwaysAvailable,
-                    $isMainLanguage,
+                    $namesByLanguage1[$languageCode],
+                    $language->id,
+                    $isMainLanguage2 && $alwaysAvailable1,
+                    $isMainLanguage2,
                     $autoLocation1['id']
                 );
-            }
-        }
-
-        foreach ($location1Entries as $row) {
-            $alwaysAvailable = (bool)($row['lang_mask'] & 1);
-            $languageIds = $this->extractLanguageIdsFromMask($row['lang_mask']);
-
-            foreach ($languageIds as $languageId) {
-                $isMainLanguage = $languageId == $location1MainLanguageId;
+            } elseif (isset($namesByLanguage2[$languageCode])) {
                 $this->internalPublishUrlAliasForLocation(
                     $location2Id,
                     $location2ParentId,
-                    $row['text'],
-                    $languageId,
-                    $isMainLanguage && $alwaysAvailable,
-                    $isMainLanguage,
+                    $namesByLanguage2[$languageCode],
+                    $language->id,
+                    $isMainLanguage1 && $alwaysAvailable2,
+                    $isMainLanguage1,
                     $autoLocation2['id']
                 );
             }
         }
     }
 
+    /**
+     * @param array $contentInfo
+     *
+     * @return array
+     */
+    private function getNamesByLanguage($contentInfo)
```

Comment:
nitpick: sounds more like a `getNamesInAllLanguages()`

Reply:
Thanks, I renamed it to `getNamesForAllLanguages()`.